File: TeamCode/src/main/java/woen239/enumerators/RunStatus.java
package woen239.enumerators;


import java.util.ArrayList;

import static org.woen.telemetry.Configs.PROCESS_ID.UNDEFINED_PROCESS_ID;

import org.woen.telemetry.ThreadedTelemetry;


public class RunStatus
{
    private final int INTEGER_MAX_VALUE = 2147483647;
    private int _currentActiveProcessId = -1;

    public enum Priority
    {
        PRIORITIZE_LOW_PROCESS_ID,
        PRIORITIZE_HIGH_PROCESS_ID,
        USE_CUSTOM_PRIORITY
    }



    private Priority _prioritySetting = Priority.PRIORITIZE_HIGH_PROCESS_ID;

    private final ArrayList<Integer> _processQueue = new ArrayList<>();
    private final ArrayList<Integer> _terminationList = new ArrayList<>();



    public RunStatus()
    {
        _prioritySetting = Priority.PRIORITIZE_HIGH_PROCESS_ID;
    }
    public RunStatus(Priority priority)
    {
        _prioritySetting = priority;
    }



    public void FullResetToActiveState()
    {
        ClearAllProcesses();
        ClearAllTerminations();
        ClearCurrentActiveProcess();
    }



    public int GetCurrentActiveProcess()
    {
        return _currentActiveProcessId;
    }
    public void ClearCurrentActiveProcess()
    {
        _currentActiveProcessId = UNDEFINED_PROCESS_ID;
    }
    public void SetCurrentActiveProcess(int processId)
    {
        _currentActiveProcessId = processId;
    }



    public int Current(int pos)
    {
        return _processQueue.get(pos);
    }
    public int size()
    {
        return _processQueue.size();
    }
    public boolean IsNotBusy()
    {
        return _processQueue.isEmpty();
    }
    public boolean IsUsedByAnyProcess()
    {
        return !_processQueue.isEmpty();
    }
    public boolean IsUsedByAnotherProcess(int ourProcessId)
    {
        for (int i = 0; i < _processQueue.size(); i++)
        {
            if (_processQueue.get(i) != ourProcessId)
                return false;
        }

        return true;
    }
    public boolean IsThisProcessTerminated(int processId)
    {
        return _terminationList.contains(processId);
    }



    public void AddProcessToQueue(int processId)
    {
        _processQueue.add(processId);
    }
    public void ClearAllProcesses()
    {
        _processQueue.clear();
    }
    public void TryToRemoveProcessIdFromQueue(int processId)
    {
        int position = _processQueue.indexOf(processId);
        while(position != -1)
        {
            position = _processQueue.indexOf(processId);
            if (position != -1) _processQueue.remove(position);
        }
    }
    public boolean IsThisProcessHighestPriority(int processId)
    {
        int maxId = 0, minId = INTEGER_MAX_VALUE;
        boolean containsRequestedId = false;

        for (int i = 0; i < _processQueue.size(); i++)
        {
            int curProcess = _processQueue.get(i);
            if (curProcess == processId)
                containsRequestedId = true;

            if (curProcess > maxId) maxId = curProcess;
            if (curProcess < minId) minId = curProcess;
        }

        return containsRequestedId
            &&
                (
                    _prioritySetting == Priority.PRIORITIZE_HIGH_PROCESS_ID
                    && processId >= maxId
                ||
                    _prioritySetting == Priority.PRIORITIZE_LOW_PROCESS_ID
                    && processId <= minId
                );
    }
    public int GetHighestPriorityProcessId()
    {
        int maxId = 0, minId = INTEGER_MAX_VALUE, processId = 0;

        for (int i = 0; i < _processQueue.size(); i++)
        {
            int curProcess = _processQueue.get(i);

            if (curProcess > maxId)
            {
                maxId = curProcess;
                processId = i;
            }
            if (curProcess < minId)
            {
                minId = curProcess;
                processId = i;
            }
        }

        return processId;
    }



    public void AddProcessToTerminationList(int processId)
    {
        TryToRemoveProcessIdFromQueue(processId);
        _terminationList.add(processId);
    }
    public boolean IsForcedToTerminateThisProcess(int processId)
    {
        for (int i = 0; i < _terminationList.size(); i++)
        {
            if (_terminationList.get(i) == processId)
                return true;
        }

        return false;
    }
    public void TryToRemoveThisProcessFromTerminationList(int processId)
    {
        int position = _terminationList.indexOf(processId);
        while(position != -1)
        {
            position = _terminationList.indexOf(processId);
            if (position != -1) _terminationList.remove(position);
        }
    }
    public void ClearAllTerminations()
    {
        _terminationList.clear();
    }



    public void ChangePrioritySetting(Priority newPrioritySetting)
    {
        _prioritySetting = newPrioritySetting;
    }
}

File: TeamCode/src/main/kotlin/org/woen/utils/process/RunStatus.kt
package org.woen.utils.process


import kotlin.math.abs
import java.util.concurrent.CopyOnWriteArrayList

import org.woen.telemetry.configs.ProcessId



class RunStatus
{
    private var _prioritySetting = Priority.PRIORITIZE_HIGH_PROCESS_ID

    private val _processQueue    = CopyOnWriteArrayList<Int>()
    private val _terminations    = CopyOnWriteArrayList<Int>()
    private var _activeProcessId = ProcessId.UNDEFINED_PROCESS_ID



    enum class Priority
    {
        PRIORITIZE_LOW_PROCESS_ID,
        PRIORITIZE_HIGH_PROCESS_ID,

        PRIORITIZE_CLOSER_TO_ZERO_FIRST_POSITIVE,
        PRIORITIZE_CLOSER_TO_ZERO_FIRST_NEGATIVE,

        PRIORITIZE_FARTHER_FROM_ZERO_FIRST_POSITIVE,
        PRIORITIZE_FARTHER_FROM_ZERO_FIRST_NEGATIVE,
    }



    constructor(prioritySetting: Priority = Priority.PRIORITIZE_HIGH_PROCESS_ID)
    {
        _prioritySetting = prioritySetting
        fullResetToActiveState()
        changePrioritySetting(prioritySetting)
    }



    fun fullResetToActiveState()
    {
        clearProcessQueue()
        clearTerminations()
        clearActiveProcess()
    }



    fun getActiveProcess() = _activeProcessId
    fun clearActiveProcess()
    {
        _activeProcessId = ProcessId.UNDEFINED_PROCESS_ID
    }
    fun setActiveProcess(processId: Int)
    {
        _activeProcessId = processId
    }



    fun isNotBusy()          =  _processQueue.isEmpty()
    fun isUsedByAnyProcess() = !_processQueue.isEmpty()
    fun isUsedByThisProcess(processId: Int)
        = _processQueue.contains(processId)
    fun isUsedByAnotherProcess(vararg exceptionProcessesId: Int): Boolean
        = _processQueue.any { it !in exceptionProcessesId }
    fun countOfThisProcess(processId: Int)
        = _processQueue.count { it == processId }



    fun addProcessToQueue(processId: Int) = _processQueue.add(processId)
    fun clearProcessQueue() = _processQueue.clear()
    fun removeProcessFromQueue(processId: Int)
        = _processQueue.removeAll { it == processId }
    fun removeOneInstanceOfProcessFromQueue(processId: Int)
        = _processQueue.remove(processId)



    fun isThisProcessHighestPriority(processId: Int,
        vararg exceptionProcessesId: Int): Boolean
    {
        if (_processQueue.isEmpty()) return false

        var maxId = Int.MIN_VALUE
        var minId = Int.MAX_VALUE
        var zeroPositive = Int.MAX_VALUE
        var zeroNegative = Int.MIN_VALUE
        var containsRequestedId = false

        for (curProcess in _processQueue)
        {
            if (curProcess == processId) containsRequestedId = true

            if (!exceptionProcessesId.contains(curProcess))
            {
                if (curProcess > maxId) maxId = curProcess
                if (curProcess < minId) minId = curProcess

                if (curProcess > 0 && curProcess < zeroPositive)
                    zeroPositive = curProcess
                if (curProcess < 0 && curProcess > zeroNegative)
                    zeroNegative = curProcess
            }
        }

        return containsRequestedId &&
                processId == returnPrioritized(maxId, minId, zeroPositive, zeroNegative)
    }
    fun getHighestPriorityProcess(vararg exceptionProcessesId: Int): Int
    {
        if (_processQueue.isEmpty()) return ProcessId.UNDEFINED_PROCESS_ID

        var maxId = Int.MIN_VALUE
        var minId = Int.MAX_VALUE
        var zeroPositive = Int.MAX_VALUE
        var zeroNegative = Int.MIN_VALUE

        for (curProcess in _processQueue)
        {
            if (!exceptionProcessesId.contains(curProcess))
            {
                if (curProcess > maxId) maxId = curProcess
                if (curProcess < minId) minId = curProcess

                if (curProcess > 0 && curProcess < zeroPositive)
                    zeroPositive = curProcess
                if (curProcess < 0 && curProcess > zeroNegative)
                    zeroNegative = curProcess
            }
        }
        return returnPrioritized(maxId, minId, zeroPositive, zeroNegative)
    }
    private fun returnPrioritized(
        maxId: Int, minId: Int,
        zeroPositive: Int, zeroNegative: Int): Int
    {
        return when (_prioritySetting)
        {
            Priority.PRIORITIZE_LOW_PROCESS_ID  -> minId
            Priority.PRIORITIZE_HIGH_PROCESS_ID -> maxId

            Priority.PRIORITIZE_CLOSER_TO_ZERO_FIRST_POSITIVE ->
                if  (zeroPositive <= abs(zeroNegative))
                    zeroPositive
                else zeroNegative
            Priority.PRIORITIZE_CLOSER_TO_ZERO_FIRST_NEGATIVE ->
                if  (abs(zeroNegative) <= zeroPositive)
                    zeroNegative
                else zeroPositive

            Priority.PRIORITIZE_FARTHER_FROM_ZERO_FIRST_POSITIVE ->
                if  (maxId >= abs(minId)) maxId
                else minId
            Priority.PRIORITIZE_FARTHER_FROM_ZERO_FIRST_NEGATIVE ->
                if  (abs(minId) >= maxId) minId
                else maxId
        }
    }



    fun isForcedToTerminateThisProcess(processId: Int)
        = _terminations.contains(processId)
    fun addProcessToTermination(processId: Int)
        = _terminations.add(processId)

    fun removeProcessFromTermination(vararg processId: Int)
        = _terminations.removeAll { it in processId }

    fun clearTerminations() = _terminations.clear()


    fun changePrioritySetting(newPrioritySetting: Priority)
    {
        _prioritySetting = newPrioritySetting
    }
}